plugins/node/opentelemetry-instrumentation-pino/src/pino.ts

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  context,
+  diag,
+  getSpan,
+  isSpanContextValid,
+  Span,
+  SpanContext,
+} from '@opentelemetry/api';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import {
+  Pino,
+  PinoInstrumentationConfig,
+  PinoLogger,
+  PinoGenLsCacheFunction,
+} from './types';
+import { VERSION } from './version';
+import * as semver from 'semver';
+
+const pinoVersions = ['>=5 <7'];
+const pino5Versions = '5.x';
+const pino6Versions = '6.x';
+
+export class PinoInstrumentation extends InstrumentationBase {
+  _pinoWriteSym: symbol | undefined;
+  _instrumentedLoggers = new Set<PinoLogger>();
+
+  constructor(config: PinoInstrumentationConfig = {}) {
+    super('@opentelemetry/instrumentation-pino', VERSION, config);
+  }
+
+  protected init() {
+    return [
+      new InstrumentationNodeModuleDefinition<Pino>(
+        'pino',
+        pinoVersions,
+        (pinoModule, version) => {
+          if (version === undefined) {
+            return pinoModule;
+          }
+
+          if (semver.satisfies(version, pino5Versions)) {
+            const proto = Object.getPrototypeOf(pinoModule());
```

Comment:
do the constructor have a side effect ? I mean we are creating a new instance here which can be problematic

Reply:
Afaik it does not, it's just a factory function: https://github.com/pinojs/pino/blob/v5.17.0/pino.js#L63 so no global state gets set.

Reply:
I think what would make more sense is to retrieve the `lib/proto` file to patch it, WDYT ?

Reply:
The thing with this is that the file hook gets called first, but in order to patch the prototype I need the value of `writeSym`. `lib/proto` gets loaded before `lib/symbols`, not sure if there is a non-hackish way to do it. Will give it some thought :slightly_smiling_face:

Reply:
Indeed, i don't think its possible then. Might be worth checking though

Reply:
Changed it to mixins patching now, should be simpler.